kola: provide internet access to qemu VM

tormath1 opened the pull request:
In this PR, we allow QEMU VMs to access Internet - it was preventing some tests to run because of this dependency.

# How to use

Run kola against tests requiring Internet access, example:
```
$ sudo ./bin/kola run -b cl -p qemu --qemu-image ...  cl.rkt.etcd3
=== RUN   cl.rkt.etcd3
qemu-img: warning: Deprecated use of backing file without explicit backing format (detected format of qcow2)
--- PASS: cl.rkt.etcd3 (27.66s)
PASS, output in _kola_temp/qemu-2021-04-13-1711-9135
```

# What's missing

- ~update readme with more information~ the user does not have to perform any manual action. They are done in https://github.com/kinvolk/mantle/pull/167/commits/3d69f85f3b033dd389f05e83fb23c6f9ce7a5e59
- [x] add a teardown method (to remove what kola created)
- [x] run all tests with `RequiresInternetAccess` flag and remove this flag

pothos commented on platform/local/dnsmasq.go:
```diff
@@ -108,6 +111,63 @@ func newInterface(s byte, i uint16) *Interface {
+// [0] is the network namespace veth
+// [1] is the host namespace veth
+func generateVethPair(pid int) []string {
+	x := pid%(31-16) + 16
```
This uses the last bits twice for both x and y. You can either divide in one case to shift the bits right or use a bitmask with `&` instead of `%`.

pothos replied:
```suggestion
	x := (pid/256)%(31-16) + 16
```
here the division approach

pothos commented on platform/local/dnsmasq.go:
```diff
@@ -108,6 +111,63 @@ func newInterface(s byte, i uint16) *Interface {
+// [1] is the host namespace veth
+func generateVethPair(pid int) []string {
+	x := pid%(31-16) + 16
+	y := pid % 255
```
```suggestion
	y := pid % 256
```
255 is still fine, or?

pothos commented on platform/local/dnsmasq.go:
```diff
@@ -147,6 +207,85 @@ func newSegment(s byte) (*Segment, error) {
+	// we first create a virtual ethernet pair in current network namespace
+	// based on the current process ID
+	pid := os.Getpid()
+	pair := generateVethPair(pid)
```
Since we don't use all bits of the PID there still can be clashes. Instead of erroring out, can we wrap this in a loop to detect a clash and retry? Another option coming to my mind is to create another global OS object like a TCP socket where the listening port is unique and has the right bit size for the values we want to have unique. With `0` as port the kernel will find a free port. E.g., using the two bytes of the TCP port means the address can be calculated as `172.(0001 and first half byte).(second half byte and third half byte).(fourth half byte and either 0000 or 0001)`.

pothos replied:
This may help: http://jodies.de/ipcalc?host=172.16.0.0&mask1=31&mask2=

pothos commented on platform/local/dnsmasq.go:
```diff
@@ -147,6 +207,85 @@ func newSegment(s byte) (*Segment, error) {
+		return nil, fmt.Errorf("unable to set up link: %w", err)
+	}
+
+	current, err := netns.GetFromName("kola-test")
```
Does this name conflict with other kola instances running on the host?

pothos commented on system/ns/enter.go:
```diff
@@ -15,32 +16,25 @@
+// NsEnter switches to a new network namespace. The returned function must be called in order
+// to restore the previous state
 func Enter(ns netns.NsHandle) (func() error, error) {
-	runtime.LockOSThread()
```
I think this has to stay: https://golang.org/pkg/runtime/#LockOSThread
Otherwise it seems that there is no guarantee in which network namespace the Go routine will be.

pothos commented on system/ns/enter.go:
```diff
@@ -52,15 +46,12 @@ func Enter(ns netns.NsHandle) (func() error, error) {
 	defer netns.Set(origns)
 
-	return netns.New()
+	return netns.NewNamed("kola-test")
```
This creates a single global namespace, or? Then we couldn't run many kola instances in parallel.

Do you think there is a way to keep the functions as they are and try to rather change where the new network namespace gets created and pass its handle along so that `.Enter()` can be used?

tormath1 replied:
It was for testing purposes - using named namespace, it's easier to `nsenter`.

pothos replied:
Ah, ok, then without it one has to use the PID: `nsenter --net -t PID`

pothos commented on platform/local/dnsmasq.go:
```diff
@@ -147,6 +207,85 @@ func newSegment(s byte) (*Segment, error) {
+		return nil, fmt.Errorf("unable to get NS from PID 1: %w", err)
+	}
+
+	if err := netns.Set(root); err != nil {
```
Seems this is not guaranteed due to how Go routines are scheduled and suspendable. Instead, the original version of `.Enter()` should be used to switch and go back.

tormath1 commented:
@pothos thanks for the review - comments have been addressed. :)

pothos commented on platform/local/dnsmasq.go:
```diff
@@ -108,6 +112,70 @@ func newInterface(s byte, i uint16) *Interface {
+func generateVethPair(seed int) []string {
+	// we basically use each single bit of the seed
+	// to generate an unique IP range
+	x := 0x10 | (seed & 0x0F)
+	y := ((seed >> 4 & 0x0F) << 4) | (seed >> 8 & 0x0F)
+	hostZ := ((seed >> 12 & 0x0F) << 4) | 0x00
+	nsZ := ((seed >> 12 & 0x0F) << 4) | 0x01
```
I get confused^^, maybe just a single shift and then masks:
```suggestion
	start := 0b10101100000100000000000000000000 // 172.16.0.0/12
	host := start | (seed <<1) | 0
	ns := host | 1
	x := host & 0xFF0000
	y := host & 0xFF00
	hostZ := host & 0xFF
	nsZ := ns & 0xFF
```

pothos commented on platform/local/dnsmasq.go:
```diff
@@ -108,6 +112,70 @@ func newInterface(s byte, i uint16) *Interface {
+	baseIP := fmt.Sprintf("172.%d.%d", x, y)
+
+	topology := make([]string, 3)
+	topology[0] = fmt.Sprintf("%s|%s", fmt.Sprintf("kola-ns-%d%d", x, y), fmt.Sprintf("%s.%d/31", baseIP, nsZ))
```
maybe `-%d%d%d` with the additional `, hostZ` to not have the same name twice

pothos commented on platform/local/dnsmasq.go:
```diff
@@ -147,6 +215,98 @@ func newSegment(s byte) (*Segment, error) {
+		return nil, fmt.Errorf("unable to listen on random port: %w", err)
+	}
+
+	defer listener.Close()
```
the `listener` needs to live on and only be closed when the network namespace is teared down

tormath1 commented:
@pothos thanks for reviewing again - comments have been addressed. PR is now ready to be fully reviewed :) If it looks good for you, I will rebase my fixups and reword some commits since logic has changed on the road.

pothos commented on platform/local/dnsmasq.go:
```diff
@@ -108,6 +118,74 @@ func newInterface(s byte, i uint16) *Interface {
+// [0] is the network namespace veth
+// [1] is the host namespace veth
+// [2] is the next hop for adding route
+func generateVethPair(seed int) []string {
```
maybe check that the seed value is <= 0xffff

pothos commented on platform/local/dnsmasq.go:
```diff
@@ -147,6 +225,120 @@ func newSegment(s byte) (*Segment, error) {
+	// we first create an unique virtual ethernet pair in current network namespace accross the multiple kola instances
+	// we use linux network random port attribution to assert the uniqueness of the IP range in order to avoid IP
+	// range clashes
+	listener, err := net.Listen("tcp", ":0")
```
This needs to be done in the host network namespace (`rootExit, err := ns.Enter(root)` and leave again)

pothos reviewed (commented):
I think we should also remove the `RequiresInternetAccess` flag and `PlatformsNoInternet` definitions from `kola/register/register.go`. Or keep the `RequiresInternetAccess` flag untouched and just remove `qemu` from `PlatformsNoInternet`.

tormath1 commented:
> I think we should also remove the `RequiresInternetAccess` flag and `PlatformsNoInternet` definitions from `kola/register/register.go`. Or keep the `RequiresInternetAccess` flag untouched and just remove `qemu` from `PlatformsNoInternet`.

@pothos I removed `RequiresInternetAccess` flag in (https://github.com/kinvolk/mantle/pull/167/commits/55ea9ee28ba247288b8adbb0e37ae886bea06956 and its parent commit) but not _all_ flag since we are hitting the issue :
```
=== RUN   cl.locksmith.cluster
--- FAIL: cl.locksmith.cluster (0.10s)
        harness.go:523: Cluster failed starting machines: rendering Container Linux config for platform "": error: platform must be specified to use templating
```

which should be tracked and addressed in another issue. It's better for now to keep this definitions then.

pothos commented:
The failing tests can be excluded with `ExcludePlatforms: []string{"qemu"},` until then

pothos commented:
`cl.locksmith.cluster` needs to be excluded

pothos commented:
Looks good, thanks!
The only thing I can think of is maybe checking if there is a `172.something` network already present like from Docker and not pick addresses in that range (maybe a retry loop). This is not a big problem but it could happen that a container IP address clashes and then the container can't access the network while kola runs…

tormath1 commented:
@pothos I added the retry logic: we provide the list of the host networks to the `generate` function then we check if the veth generated host IP is in one of this networks.

pothos commented on platform/local/dnsmasq.go:
```diff
@@ -108,6 +129,84 @@ func newInterface(s byte, i uint16) *Interface {
+// [0] is the network namespace veth
+// [1] is the host namespace veth
+// [2] is the next hop for adding route
+func generateVethPair(seed int, hostsIP []*net.IPNet) ([]string, error) {
```
```suggestion
func generateVethPair(seed int, hostsNetworks []*net.IPNet) ([]string, error) {
```

pothos commented on platform/local/dnsmasq.go:
```diff
@@ -108,6 +129,84 @@ func newInterface(s byte, i uint16) *Interface {
+	binary.BigEndian.PutUint32(hostIP, uint32(host))
+	binary.BigEndian.PutUint32(nsIP, uint32(ns))
+
+	for _, host := range hostsIP {
```
```suggestion
	for _, network := range hostsNetworks {
```

pothos commented on platform/local/dnsmasq.go:
```diff
@@ -108,6 +129,84 @@ func newInterface(s byte, i uint16) *Interface {
+	binary.BigEndian.PutUint32(nsIP, uint32(ns))
+
+	for _, host := range hostsIP {
+		if host.Contains(hostIP) {
```
```suggestion
		if network.Contains(hostIP) {
```

pothos commented on platform/local/dnsmasq.go:
```diff
@@ -147,6 +249,175 @@ func newSegment(s byte) (*Segment, error) {
+		pair, err = generateVethPair(p, ips)
+		if err != nil {
+			if errors.Is(err, ErrIPClash) {
+				isValid = false
```
```suggestion
				isValid = false
				plog.Debugf("failed to use port %v as unique seed for the veth pair due to address clash, retrying", p)
```
